### PR TITLE
image: build dev and release image for all arch

### PR DESCRIPTION
**Pull request by huoqifeng**
Fixed: #1521

### CODE REVIEW

File: .github/workflows/caa_build_and_push.yaml [hunk 1]
```diff
@@ -29,7 +29,7 @@ on:
         required: false
         type: string
       git_ref:
-        default: 'main'
+        default: 'image-action'
```
Comment (liudalibj) [inline .github/workflows/caa_build_and_push.yaml:32]: should revoke this change if we want to merge main?

**Review by liudalibj — APPROVED** [review]
I checked the image build works well now.
one comment, but all other changes LGTM.
Thanks @huoqifeng

File: .github/workflows/caa_build_and_push.yaml (hunk 1 above)
Comment (huoqifeng) [reply .github/workflows/caa_build_and_push.yaml]: @liudalibj yes, I'm going to rollback it. thanks!

**Comment by huoqifeng** [comment]
I was also thinking alternative:
- build and push the arch specific image to quay directly but not build the tar file.
- Pull the arch specific image in stage 2 and create/push the manifest.

**Comment by huoqifeng** [comment]
I'll start working on the alternative action: basic flow is:
- Job 1:
  - Build arch specific image and push it
  - Export the `dev_tags` and `release_tags` and upload it to GHA artifactory
  - Export the arches file to artifactory
- Job 2
  - Download the files from artifactory
  - Read `dev_tags` and `release_tags`
  - Read all arches
  - Iterate the arches and tags and generate the manifest
  - Push the manifest
- Job 3
  - Download the files from artifactory
  - Read `dev_tags` and `release_tags`
  - Read all arches
  - Verify all arches by pulling manifest from registry

**Comment by huoqifeng** [comment]
Evidence in GHA: https://github.com/confidential-containers/cloud-api-adaptor/actions/runs/6609105060
The built image: 
image

**Review by genjuro214 — APPROVED** [review]
LGTM

File: hack/image-manifest.sh [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+    print_all_arches
+}
+
+function get_dev_tags() {
+    dev_tags_array=()
+
+    local dev_tags=""
+
+	pushd "${script_dir}/.."
```
Comment (liudalibj) [inline hack/image-manifest.sh:41]: bad indent?

File: hack/image-manifest.sh [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        popd
+        exit 99
+    fi
+
+    IFS=',' read -ra dev_tags_array <<< "$dev_tags"
+    for i in "${dev_tags_array[@]}"; do
+      echo $i
+    done
+
+	popd
```
Comment (liudalibj) [inline hack/image-manifest.sh:56]: bad indent?

File: hack/image-manifest.sh [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+    print_dev_tags
+}
+
+function get_release_tags() {
+    release_tags_array=()
+
+    local release_tags=""
+
+	pushd "${script_dir}/.."
```
Comment (liudalibj) [inline hack/image-manifest.sh:66]: bad indent?

File: hack/image-manifest.sh [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        popd
+        exit 99
+    fi
+
+    IFS=',' read -ra release_tags_array <<< "$release_tags"
+    for i in "${release_tags_array[@]}"; do
+      echo $i
+    done
+
+	popd
```
Comment (liudalibj) [inline hack/image-manifest.sh:81]: bad indent?

**Review by liudalibj — APPROVED** [review]
the are some bad indents(should be blankspaces in shell)?
Others LGTM.

File: hack/image-manifest.sh (hunk 2 above)
Comment (huoqifeng) [reply hack/image-manifest.sh]: fixed, thanks @liudalibj

File: hack/image-manifest.sh (hunk 4 above)
Comment (huoqifeng) [reply hack/image-manifest.sh]: fixed

File: hack/image-manifest.sh (hunk 5 above)
Comment (huoqifeng) [reply hack/image-manifest.sh:69]: fixed

File: Dockerfile [hunk 6]
```diff
@@ -2,12 +2,7 @@ ARG BUILD_TYPE=dev
 ARG BUILDER_BASE=quay.io/confidential-containers/golang-fedora:1.20.8-36
 ARG BASE=registry.fedoraproject.org/fedora:38
 
-# This dockerfile uses Go cross-compilation to build the binary,
-# we build on the host platform ($BUILDPLATFORM) and then copy the
-# binary into the container image of the target platform ($TARGETPLATFORM)
-# that was specified with --platform. For more details see:
-# https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/
-FROM --platform=$BUILDPLATFORM $BUILDER_BASE as builder-release
+FROM --platform=$TARGETPLATFORM $BUILDER_BASE as builder-release
```
Comment (katexochen) [inline Dockerfile:5]: What is the reason for this change? This will make builds really slow, as it does not use Go cross compilation but qemu emulation of the builder.

Comment (huoqifeng) [reply Dockerfile:5]: @katexochen this is to enable multi-arch dev-* build to include libvirt options. Maybe we can split it to dev-* and release here.

Comment (katexochen) [reply Dockerfile:5]: You just need to set `supported_arches` in `--platform "${supported_arches}"` in the docker call to the platform you want to build.

Comment (huoqifeng) [reply Dockerfile:5]: @katexochen when `CGO=1` set (which is the case when libvirt enabled), I think we need use `TARGETPLATFORM` builder to build the right image. I did some trying in this PR https://github.com/confidential-containers/cloud-api-adaptor/pull/1439 before.
